usbwallet/hub.go

```diff
@@ -0,0 +1,231 @@
+	// Don't scan the USB like crazy it the user fetches wallets in a loop
+	hub.stateLock.RLock()
+	elapsed := time.Since(hub.refreshed)
+	hub.stateLock.RUnlock()
+
+	if elapsed < refreshThrottling {
+		return
+	}
+
+	// If USB enumeration is continually failing, don't keep trying indefinitely
+	if atomic.LoadUint32(&hub.enumFails) > 2 {
+		return
+	}
+
+	// Retrieve the current list of USB wallet devices
+	var devices []usb.DeviceInfo
+
+	if onLinux {
+		// hidapi on Linux opens the device during enumeration to retrieve some infos,
+		// breaking the Ledger protocol if that is waiting for user confirmation. This
+		// is a bug acknowledged at Ledger, but it won't be fixed on old devices, so we
+		// need to prevent concurrent comms ourselves. The more elegant solution would
+		// be to ditch enumeration in favor of hotplug events, but that don't work yet
+		// on Windows so if we need to hack it anyway, this is more elegant for now.
+		hub.commsLock.Lock()
+		if hub.commsPend > 0 { // A confirmation is pending, don't refresh
+			hub.commsLock.Unlock()
+			return
+		}
+	}
+	infos := usb.Enumerate(hub.vendorID, 0)
+	if infos == nil {
+		if onLinux {
+			// See rationale before the enumeration why this is needed and only on Linux.
+			hub.commsLock.Unlock()
+		}
+		return
+	}
+	atomic.StoreUint32(&hub.enumFails, 0)
+
+	for _, info := range infos {
+		for _, id := range hub.productIDs {
+			// Windows and macOS use UsageID matching, Linux uses Interface matching
+			if info.ProductID == id && (info.UsagePage == hub.usageID || info.Interface == hub.endpointID) {
+				devices = append(devices, info)
+				break
+			}
+		}
+	}
+
+	if onLinux {
+		// See rationale before the enumeration why this is needed and only on Linux.
+		hub.commsLock.Unlock()
+	}
+
+	// Transform the current list of wallets into the new one
+	hub.stateLock.Lock()
+
+	wallets := make([]accounts.Wallet, 0, len(devices))
+
+	for _, device := range devices {
+		url := gethaccounts.URL{
+			Scheme: hub.scheme,
+			Path:   device.Path,
+		}
+
+		// Drop wallets in front of the next device or those that failed for some reason
+		for len(hub.wallets) > 0 {
+			// Abort if we're past the current device and found an operational one
+			_, err := hub.wallets[0].Status()
+			if hub.wallets[0].URL().Cmp(url) >= 0 || err == nil {
+				break
+			}
+			// Drop the stale and failed devices
+			hub.wallets = hub.wallets[1:]
+		}
+
+		// If there are no more wallets or the device is before the next, wrap new wallet
+		if len(hub.wallets) == 0 || hub.wallets[0].URL().Cmp(url) > 0 {
+			wallet := &wallet{
+				hub:    hub,
+				driver: hub.makeDriver(),
+				url:    &url,
+				info:   device,
+			}
+
+			wallets = append(wallets, wallet)
+			continue
+		}
+		// If the device is the same as the first wallet, keep it
+		if hub.wallets[0].URL().Cmp(url) == 0 {
+			wallets = append(wallets, hub.wallets[0])
+			hub.wallets = hub.wallets[1:]
+			continue
+		}
+	}
+
+	hub.refreshed = time.Now()
```

Comment:
```suggestion
	// #nosec
	hub.refreshed = time.Now().UTC()
```

Reply:
@fedekunze This is not supported by CodeQL (see [this discussion](https://github.com/github/codeql/issues/9298)) - `nosec` is for the [Gosec](https://github.com/securego/gosec) linter warnings.